src/Diagnostics/Debug/StateCheck.jl

```diff
@@ -0,0 +1,556 @@
+               Created for regression testing and code change tracking and debugging.
+               StateCheck functions iterate over named variables in an MPIStateArray,
+               calculate and report their statistics and/or write values for all or
+               some subset of points at a fixed frequency.
+
+               Functions:
+               StateCheck.sccreate()   - Create a StateCheck call back variable.
+
+               StateCheck.scdocheck()  - Check Statecheck variable values against reference values.
+
+               StateCheck.scprintref() - Print Statecheck variable in format for creating 
+                                         reference values.
+"""
+module StateCheck
+
+# Imports from standard Julia packages
+using Formatting
+using MPI
+using Printf
+using StaticArrays
+using Statistics
+
+# Imports from ClimateMachine core
+import ClimateMachine.GenericCallbacks: EveryXSimulationSteps
+import ClimateMachine.MPIStateArrays: MPIStateArray
+import ClimateMachine.VariableTemplates: flattenednames
+
+####
+# For testing put a new function signature here!
+# Needs to go in src/Utilities/VariableTemplates/var_names.jl
+# This handles SMatrix case
+flattenednames(::Type{T}; prefix = "") where {T <: SArray} =
+    ntuple(i -> "$prefix[$i]", length(T))
+####
+
+struct vstat
+    max
+    min
+    mean
+    std
+end
+
+# Global functions to expose
+# sccreate - Create a state checker call back
+export sccreate
+export scdocheck
+export scprintref
+
+# ntFreqDef:: default frequency (in time steps) for output.
+# precDef  :: default precision used for formatted output table
+ntFreqDef = 10
+precDef = 15
+
+"""
+ sccreate :: Create a a "state check" call-back for one or more MPIStateArrays  \n
+             that will report basic statistics for the fields in the array.
+
+             Input:
+                fields: A required first argument that is an array of one or more
+                        MPIStateArrays and label string pair tuples.
+                        State array statistics will be reported for the named symbols
+                        in each MPIStateArray labeled with the label string.
+                ntFreq: An optional second argument with default value of 
+                        $ntFreqDef that sets how freuently (in time-step counts) the
+                        statistics are reported.
+             Named:
+                  prec: A named argument that sets number of decimal places to print for
+                        statistics, defaults to $precDef.
+
+             Return: 
+                sccb: A state checker that can be used in a callback().
+
+             Example:
+             julia> using ClimateMachine.VariableTemplates
+             julia> using StaticArrays
+             julia> using ClimateMachine.MPIStateArrays
+             julia> using MPI
+             julia> MPI.Init()
+             julia> T=Float64
```

Comment:
```suggestion
             julia> FT=Float64
```
We've been using `FT` to indicate the float type. Would be nice to keep this consistent (a big part to avoid ambiguity with temperature `T`)